### PR TITLE
Windows HDR support "option two"

### PR DESCRIPTION
**Pull request by thexai**
## Description
Inspired on work of @fandangos, this implementation offers same HDR functionality **without using proprietary NVIDIA API.**

## Motivation and Context
Enable Windows HDR support on KODI in a stable and compatible way.

## How Has This Been Tested?

- To use HDR features enable Windows HDR switch before open Kodi or toggle HDR ON from power options menu in Kodi.
- If video has HDR metadata enables proper rendering.
- Can also play SDR content with Windows HDR switch ON and it is displayed correctly.
- And HDR content played with Windows HDR switch OFF is displayed correctly too (with no HDR).
- Only uses Microsoft Windows standard API to HDR passthrough.
- If Windows HDR switch is turned OFF, it behaves like the normal Kodi edition.
- For devices without HDR display capable uses HW tone mapping (DXVA) or SW tone mapping (pixel shaders) for HDR to SDR conversion.

Tested on Intel NUC 8i3BEK + Dennon AVX-1600H + Sony TV KD-55AG9 OLED.

Tested with DXVA render:
HDR content ----> HDR ON = OK native
SDR content -----> HDR ON = OK 10 bit swapchain can handle 709 color too
HDR content -----> HDR OFF = OK with HW tonemapping
SDR content -----> HDR OFF = OK native

Tested with pixel shaders render:
HDR content ----> HDR ON = OK native
SDR content -----> HDR ON = OK 10 bit swapchain can handle 709 color too
HDR content -----> HDR OFF = OK with SW tonemapping
SDR content -----> HDR OFF = OK native

Also works on NVIDIA / AMD hardware.

Public releases: https://forum.kodi.tv/showthread.php?tid=349861

## Screenshots (if appropriate):
System HDR capable with Windows HDR off
HDR_on

System HDR capable with Windows HDR on
HDR_off

System no HDR capable (normal display)
HDR_no-capable

## Types of change

- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by fuzzard** [comment]
All the petty shit aside, thats more inline with what i was suggesting.

When you closed the last PR, i took some time to implement what i was suggesting to you

https://github.com/fuzzard/xbmc/tree/Windows-HDR-refactored

Probably the main differences i went with that are still there

1) Remove the application.cpp logging. Doesnt really serve a benefit, and as you said, that logging block is done extremely early in startup. I just dont think it brings any benefit.

2) IsHDRDisplay, you check for 3 states. i changed that to a boolean to check if external display is hdr capable. its a binary state. Makes use of existing override. You're 3rd state is about an OS level setting being on/off. Keep it separate.

3) i introduced GetOSHDRStatus into base class. The whole OS toggle thing is pretty much unique to windows. My reasoning, and honestly it doesnt really sit well with me still, was Apple TV could potentially use the same check. Its a reach, but it makes SystemGuiInfo non platform specific.
To be honest, now i think about it, its probably just best to ifdef windows there and be done with it.

4) changed your gui/skinning to a single toggle.

5) this would be the biggest change in my commit, and it was just reworking what is int CWIN32Util::GetOSHDRStatus() and bool CWIN32Util::SetHDR() ( is called GetHDRDisplayStatus and CWIN32Util::ToggleWindowsHDR() in your code). Id like to refactor them both, as there is a lot of common code, but that would have been for another day.

I like to exit early. Removes the overly large indentations of if blocks. 

I also wanted to look at how/where you're putting the methods. scattered between utils and deviceresources. I have a feeling i would have relocated some of those methods, but thats a super minor thing, and havent really put any thought into it to be able to give any suggestion/reasoning for it.

Feel free to take/disregard anything in my commit. I dont have a windows box at all to test on, so thats pretty much where i ended up for the day.

**Comment by thexai** [comment]
> Remove the application.cpp logging. Doesnt really serve a benefit, and as you said, that logging block is done extremely early in startup. I just dont think it brings any benefit.

With the current implementation it doesn't bother to leave it there. I think it can be useful to see quickly in the logs if it is a version of kodi that supports HDR.

> i introduced GetOSHDRStatus into base class. The whole OS toggle thing is pretty much unique to windows. My reasoning, and honestly it doesnt really sit well with me still, was Apple TV could potentially use the same check. Its a reach, but it makes SystemGuiInfo non platform specific.
To be honest, now i think about it, its probably just best to ifdef windows there and be done with it.

It is good to know that a new function can be introduced. I would take the opportunity to return 3 states and with a single function to obtain the status of the display as well (in reality it is the same) since it is obtained in the same way.

`int GetOSHDRStatus();`

Otherwise, two consecutive calls to the same (similar / shared code) function will be necessary. I want to avoid calling the same code twice.

> changed your gui/skinning to a single toggle.

I have to study how it works. I imagine it is possible to change the text that appears on the button.

"Toggle display HDR on"  and "Toggle display HDR off" in function of a variable. 

> this would be the biggest change in my commit, and it was just reworking what is int CWIN32Util::GetOSHDRStatus() and bool CWIN32Util::SetHDR() ( is called GetHDRDisplayStatus and CWIN32Util::ToggleWindowsHDR() in your code). Id like to refactor them both, as there is a lot of common code,

I agree. But this can be left for later. I would like to define and set the windowing/base class methods first.

**Comment by thexai** [comment]
Updated screenshots to clarify how it works power options menu.

**Comment by da-anda** [comment]
from what I read on the forum, the toggle feature currently closes KODI and one has to manually start it again. So from this POV having the buttons in would only cause lots of support questions (like it was the case in the forum post already -> "KODI crashes when I click the toggle"). I think my take on this would be to not have the buttons in the initial commit and add them afterwards. So the focus for now should be to get the architecture right and HDR working and add usability improvements afterwards. But that's just my 2c.

> It is good to know that a new method can be introduced into base class. I would take the opportunity to return 3 states and with a single function to obtain the status of the display as well (in reality it is the same) since it is obtained in the same way.
>
>int GetOSHDRStatus();
>
>Otherwise, two consecutive calls to the same (similar / shared code) function will be necessary. I want to avoid calling the same code twice.
>
>e.g.: first call IsHDRDisplay second call GetOSHDRStatus and two functions call same code in CWIN32Util.

can't you just cache the result somewhere in windows specific code? As it has been mentioned, we are looking for an implementation/architecture that makes sense for all platforms and is future proof, so from our POV splitting the methods would make more sense.

**Comment by thexai** [comment]
> from what I read on the forum, the toggle feature currently closes KODI and one has to manually start it again. So from this POV having the buttons in would only cause lots of support questions (like it was the case in the forum post already -> "KODI crashes when I click the toggle"). 

It's implemented in last commit (toggle HDR/SDR without restart)   ;-)

And many more changes. Read commit log!

I hope to go on the right track...

**Comment by thexai** [comment]
> IsHDRDisplay, you check for 3 states. i changed that to a boolean to check if external display is hdr capable. its a binary state. Makes use of existing override. You're 3rd state is about an OS level setting being on/off. Keep it separate.

> i introduced GetOSHDRStatus into base class. The whole OS toggle thing is pretty much unique to windows. My reasoning, and honestly it doesnt really sit well with me still, was Apple TV could potentially use the same check. Its a reach, but it makes SystemGuiInfo non platform specific.

Implemented as you ask. Now is bool.  https://github.com/xbmc/xbmc/pull/17106/commits/a77557d66998db9988d86eb26c401aae48b784a8

**Comment by da-anda** [comment]
cool that you were able to toggle the HDR mode without requiring a restart of KODI. My question now would be, does it in this case still make sense to have a manual toggle in the power menu or wouldn't it be better if we could automatically trigger the HDR mode when playback starts? For stereoscopic playback (3D) we also do it automatically and have an additional toggle in the fullscreen video playback OSD. Same might probably make sense for HDR as well - at least on the long run.

**Comment by thexai** [comment]
> wouldn't it be better if we could automatically trigger the HDR mode when playback starts?

It's no easy possible because display switch overrides refresh rate match.
Can be switch before but before of whitelist match metadata is not yet demuxed. It's not trivial.

**Comment by thexai** [comment]
Jenkins build this please

**Comment by thexai** [comment]
A forum user @_DMU_ with HDMI analyzer confirms now that HDR metadata passthrough is perfect:

**after:**
Kodi

**now:**
Kodi_2

https://forum.kodi.tv/showthread.php?tid=345566&pid=2914163#pid2914163

**Comment by thexai** [comment]
Jenkins build this please

File: xbmc/windowing/win10/WinSystemWin10DX.cpp [hunk 1]
```diff
@@ -152,3 +153,47 @@ void CWinSystemWin10DX::UninitHooks()
...
 {
 }
+
+bool CWinSystemWin10DX::SetHDR(const VideoPicture* videoPicture /*not used*/)
+{
+  bool success = CWIN32Util::ToggleWindowsHDR();
+
+  if (success)
+  {
+    Sleep(2000); //Display is switching
```
Comment (da-anda) [inline xbmc/windowing/win10/WinSystemWin10DX.cpp]: do you think it would make sense here to use the user configurable "refresh rate switch delay" here instead of the fixed 2 seconds? I suppose that HDR switching is basically taking as long as refresh rate switching, but that's just a guess.

**Comment by Paxxi** [comment]
Hi @thexai I'm one of the Windows platform developers and I will try and review this so that we hopefully can get it merged for v19. Normally afedchin is the expert in this area but he's busy with work so I'm trying to fill in but I'm having to get up to speed on how HDR works and specifically in Windows.

I just wanted to let you know that I'm working on reviewing this and it's not forgotten about. I'll get back to you as soon as possible with more specific feedback.

I did try the changes on an xbox one and best I can tell it looked good and worked fine.

For now, thank you for all your work and your patience 😄

**Comment by thexai** [comment]
Tested today HDR in UWP (win10) and works perfectly.

However HDR needs to be activated in Windows display settings. Toggle inside Kodi not works because a bug in my code or limitation of UWP API.... needs to be investigated.

**Comment by thexai** [comment]
Jenkins build this please

**Comment by Paxxi** [comment]
Ok I have been thinking about this a lot over the last week and my current opinion is that we should focus on the HDR playback for now and drop the HDR os switching.

The HDR os switching is questionable if it's a good idea. There's a reason that Microsoft chose to not offer an easy API for the HDR toggle and leaves it up to the user. This is a global setting that affects other software as well and if several apps try to toggle this on/off things might end up in all sorts of broken states.
There's also the fact that we can't do this for UWP and it makes sense to keep the behavior consistent to make it easier for users and support.

The deciding factor for me is that the HDR os switching can be implemented as a python addon so by not having it in Kodi core we aren't blocking this possibility but we keep core simple and more consistent across platforms.

We should be able to use [this](https://github.com/microsoft/DirectX-Graphics-Samples/blob/29fb35649a449da30adcbfefee5d2e1a8258c718/Samples/Desktop/D3D12HDR/src/D3D12HDR.cpp#L1114) method to check if the system is HDR capable, and put that in some method we already have that reacts to changes in the window system like `DX::DeviceResources::HandleOutputChange`. I haven't been able to test that this method works for all scenarios so there could be a better way to handle it.

I think we should set static HDR metadata if we don't find any dynamic metadata in the stream, maybe set this directly at start as it would be overwritten later by dynamic metadata anyway?

@fandangos mentioned on slack that refresh rate switching doesn't work when HDR is enabled so this is something that needs to be investigated so we don't lose that functionality.

In summary
1. Remove everything for OS HDR switching, win32util, skin, powermenu etc
2. Do the HDR capability check in an event such as `HandleOutputChanged`
3. Make sure that refresh rate is switched correctly to match the video frame rate.
4. optional, use static metadata by default or if no dynamic metadata is found.

Thank you for all your work. With the above changes I'm fairly confident that we can get this merged and have HDR support in v19.

**Comment by thexai** [comment]
Ok but people won't use it.

People complain even because they have to manual switch with the button... they want it to be fully automatic.

They will continue to use the fork because it is more practical.


> @fandangos mentioned on slack that refresh rate switching doesn't work when HDR is enabled so this is something that needs to be investigated so we don't lose that functionality.

This happens on his fork. In mine it doesn't happen. The current manual switch option does not interfere with refresh rate match.

**Comment by fandangos** [comment]
I just mentioned with Paxxi that I poorly explained it to him. 

If the function to turn hdr on/off is in CRenderBase::Configure it will break refresh rate switching. That's one of the reasons why I deleted my entire second fork of this.

For an automatic solution we are kinda of screwed:
1. Automatic method before video starts: doesn't have the metadata to check and be started or not.
2. Automatic method during Configure: breaks refresh rate switching.
3. Automatic method after video started: wrong metadata.

The idea of autohdr is good for nvidia and amd gpus but as we discussed on the forum doesn't work for ryzen and Intel.
Since the code to switch hdr on uses unlisted and private microsoft apis that are not documented (as far as I know) it would remove a burden of the code if this was dropped and it could surface as an independent addon in the future.

This would make it safer.
Also people will stop using my fork because as I explained in the thread, I won't be updating it any further since this will probably get merged.

Just my thoughts on this.
Thank you for all your work, Jogal.

**Comment by thexai** [comment]
> We should be able to use this method to check if the system is HDR capable,

`ComPtr<IDXGIOutput6> output6;`   needs to be compiled with  `<dxgi1_6.h>`  this breaks runtime compatibility with Windows 7

**Comment by dvidebaek** [comment]
Why dont we stop supporting Windows 7 if it is Holding u’s back? Like Python 2 it is time to move on

**Comment by Paxxi** [comment]
I thought that it would still run on windows 7 but the `As` call would fail and return nullptr when used like the example

@dvidebaek I agree that it's about time to drop support for Windows 7 now that it's out of support but that's a larger decision and if we can keep support with minimal effort it's worth it for v19.

**Comment by dvidebaek** [comment]
Kodi is in need of more Windows developers. Why not say that Leia 18.x will be the last officially supported Kodi for Windows 7. Then the few Windows developers could use there free spare time on implementing new features insted of supporting a legacy OS... HDR is the future not Windows 7. Windows 7 is a waste of time supporting.

**Comment by thexai** [comment]
We are making it very very difficult for me. If I close this PR I will not open any more...

- As it is now it works.
- It does not break compatibility with Windows 7.
- The auto refresh rate match works.
- Gives the possibility of activating the HDR mode without leaving to OS control panel. Remove this is not permissible if Kodi is used with remote control.
- Image quality and HDR metadata passthrough is perfect.
- Works in UWP with some limitations due UWP API  but works with same image quality.

**Comment by thexai** [comment]
> Since the code to switch hdr on uses unlisted and private microsoft apis that are not documented (as far as I know) it would remove a burden of the code if this was dropped and it could surface as an independent addon in the future.

I'm sorry but this is not true.
Of everything that is said in the forums there are many things that are not true.

Anotación 2020-01-12 162004

Anotación 2020-01-12 161915

https://docs.microsoft.com/en-us/windows/win32/api/wingdi/ne-wingdi-displayconfig_device_info_type

**Comment by fandangos** [comment]
As I've said, as far as I know. 

I have worked with private APIs for HDR, not windows api. 

When I'm not sure I add the 'as far as I know' before commenting since, as I've said many times before, I'm still learning.

Also I base myself on the readme that came along with that code that you and I copied. Have you read it? It explains some of that function.

The author, 'AppWizard', claimed it was undocumented function calls. You can find the original files in the hdr thread, _DMU_ was the one offering those in a google drive link.

So who am I to contradict a readme from a code that I partially understand? :)

And finally this is not my call, it's team kodi's. All of them have it very clear I'm just a beginner in this.
So if I say something that is incorrect, please, feel free to correct it, as you are the more experienced one.

**Comment by thexai** [comment]
I agree but. "It's undocumented" or "not understand it" is not a valid reason to remove it.

**Comment by dvidebaek** [comment]
Dont stop @thexai you are the Windows communities only hope for bringing HDR to Windows. I know you can figure out how to implement automatic HDR on/off. Keep going. I have a feeling TK only will work on HDR when the Linux Kernel is ready. It could take months!!!

**Comment by thexai** [comment]
> The idea of autohdr is good for nvidia and amd gpus

This is also not true:

It's causing issues with refresh rate match beacause:

AUTOHDR.exe switch display to HDR mode and NOT waits (safe) 2 seconds to complete.

Due this Kodi opens in an indeterminate state in SDR swapchain mode but display in HDR mode.

When playback begins and calls ResizeBuffers causes issues with refresh rate match because swapchain configuration changes from SDR to HDR in an uncontrolled way.

These issues are NOT PRESENT if is not used AUTOHDR.exe

People uses AUTOHDR.exe because YouTube video is cool……....

People complain about refresh rate issues with HDR........

**Comment by fandangos** [comment]
No issues for me and some others who used it. 

I made it before you added the advancedsettings function to your fork and I even added on the first post that no one would need to use it. It was just an option.

If adding a Sleep(2000); to it fix the problem I can do it when I get home, no problem. Thanks for reporting it.

And the youtube video was really cool, glad you liked it. Thank you. :))

Well, either way, if you feel like I'm interfering or causing any problems to your work I'll ask the kodi mods to delete the autohdr thread.
I already stopped working on my fork, no big deal deleting autohdr. I really don't want to get in the way.
Just let me know.

**Comment by da-anda** [comment]
> I have a feeling TK only will work on HDR when the Linux Kernel is ready. It could take months!!!

Like with every feature in KODI it will be worked on once any dev (regardless if team member or not) is interested in working on it. TK devs are no different from community devs, other than that at some point they got invited as a sign of gratitude for their continuous work on KODI. There are no additional obligations or attached strings in becoming a team member - devs keep all their individual freedom and are free to do whatever they want.
And yes, given that we have more Linux interested devs than Windows, Mac or Android, chances are higher to get HDR there more quickly once possible there. But all it takes is one interested dev, and we had that for Android and it's working there already for some time now (albeit differently since Android is a bypass renderer).

@thexai are you still tweaking on it, or are you waiting on a response from us?

**Comment by thexai** [comment]
My POV is clear:

I do not want to remove the code for manual HDR / SDR switch because this removes important functionality and usability.

Still I consider that the menu in power settings is a temporary solution.

I also consider that because Kodi 19 is in alpha state it could be merged like this and continue improving later (I am the first interested because I use it almost every day).

For my part, I would continue working on it (automatic switching) when I have free time…

My top priorities are:

1.	Image quality in SDR and HDR
2.	Stability
3.	Compatibility with non-HDR systems (Kodi works as before for these users)
4.	Compatibility with other platforms (which does not interfere at the compilation level and does not break anything)
5.	User friendly experience / GUI / skinning / addons

If it is not merged because is an “imperfect” / incomplete job, this job is lost (lost by Kodi users/community.  I will continue to use it in my home).

**Comment by da-anda** [comment]
> I do not want to remove the code for manual HDR / SDR switch because this removes important functionality and usability.

@thexai we get that this is an important feature to improve usability, but you admitted yourself that this can only be a temporary solution. And it our eyes it is quite a hack and moved to a place where it a) doesn't really belong and b) is up to the skin to actually add it there, which again is not ideal for a feature you consider essential.

The suggestion of paxxi (and the preference of the team) was, and still is, to move the trigger for this feature to an add-on, making use of some utility `exe` (like the one of @fandangos) that performs the switch of the HDR mode of the OS.

The benefit of moving the trigger logic into an add-on would be that users have most flexibility on when it should be triggered. It could be on startup of KODI, via context menu entry on HDR movies in the library, as soon as one enters the video section of his library, on click of a remote button and what not. So it would be much more configurable and adjustable to the users needs/wishes and in addition would keep the core clean for now and give us more time to find a better solution.

So I'm sorry to say, if you want this PR to get merged, the HDR switching logic has to be moved to an add-on.

PS: if it helps in some way, since Win7 is EOL now, we will also drop Win7 support soon. For now it's planned to happen with v20, but if HDR would benefit of APIs that would break backwards compatibility to v7, we would consider dropping it for v19 already. We are still discussing this ofc - so decision is not final yet.

**Comment by thexai** [comment]
> to move the trigger for this feature to an add-on, making use of some utility exe (like the one of @fandangos) that performs the switch of the HDR mode in the OS.
> 
> The benefit of moving the trigger logic into an add-on would be that users have most flexibility on when it should be triggered. It could be on startup of KODI, via context menu entry on HDR movies in the library, as soon as one enters the video section of his library, on click of a remote button and what not. So it would be much more configurable and adjustable to the users needs/wishes and in addition would keep the core clean for now and give us more time to find a better solution.
> 
> So I'm sorry to say, if you want this PR to get merged, the HDR switching logic has to be moved to an add-on for the time being.

I'm sorry but it's a very stupid idea. I will not waste my time explaining why.

Some clue:

- Python needs to call a Win32 API code "tool.exe" that is already in Kodi.exe
- Kodi GUI needs know if it should show the button for HDR switch (needs know if the system is HDR capable), then needs part of the code that is supposed to be implemented outside.
- "Current GUI settings / Player / Use HDR display capabilities" needs know if system supports HDR to shown setting.
- Windows HDR switch from outside OK..... but swapchain is reconfigured by Python hahahhah

**Comment by Paxxi** [comment]
We have tried to be accommodating and clearly stated which parts we find troublesome currently and given you a path to get this merged .

Nobody said the swapchain would be configured by python, only that the code in Win32Util can be in an addon if it's needed.

Neither me nor @da-anda said that you used undocumented apis. I have some concerns over the usage of the DisplayConfiguration apis because the documentationen around them is very sparse compared to what win32 apis usually have available. I'm not an expert in the area so I gave you the option to remove that part to get your work merged.

You continue to be arrogant and belittling despite our efforts trying to work with you in good faith. You're very capable technically as shown by the work you've done here but the way you treat people is not acceptable.

This PR is done and it won't get approved. You're welcome to try again in the future if you can learn to behave better.

**Comment by dvidebaek** [comment]
Sad to see this. Thanks for the effort @thexai

**Comment by da-anda** [comment]
@thexai 

>    Kodi GUI needs know if it should show the button for HDR switch (needs know if the system is HDR capable), then needs part of the code that is supposed to be implemented outside.

no, KODI does not have to know if this button has to be shown. Since the add-on is responsible for switching the OS HDR mode, it is also responsible to inject the button or toggle it automatically. The possibilities are almost endless and up to the add-on logic where to inject a context menu item or toggle automatically.

>    "Current GUI settings / Player / Use HDR display capabilities" needs know if system supports HDR to shown setting.

And it still can. Nobody said that you have to remove display capability enumeration in KODI etc. - just the power menu button and the logic to change the system setting.

> Windows HDR switch from outside OK..... but swapchain is reconfigured by Python hahahhah

nobody said that everything had to move into an add-on - only the logic to flip the HDR switch in the OS. Yes, that would mean that some logic might have to be duplicated in the tool (HDR capability detection) or that add-ons might have to be able to read some system information via APIs from KODI. In fact, there are already ways to get system info from within the Python API or JSON-RPC API. Those APIs could be extended to provide additional info and maybe even a method to enable/disable KODIs HDR mode if this would be needed (like it's possible to trigger the stereoscopic 3D mode from python and JSON-RPC).
So there would be ways to do what we asked for, but you completely shut down and don't seem to be open to compromise. That's really sad to see.

@dvidebaek we do want HDR, just not the power menu switch and the logic to change the OS setting. Everything else was fine (with room for later improvement), but thexai unfortunately does not seem to be willing to find a compromise to outsource this settings toggle or communicate with us in a cooperative way.

**Comment by dvidebaek** [comment]
> @thexai
> 
> > Kodi GUI needs know if it should show the button for HDR switch (needs know if the system is HDR capable), then needs part of the code that is supposed to be implemented outside.
> 
> no, KODI does not have to know if this button has to be shown. Since the add-on is responsible for switching the OS HDR mode, it is also responsible to inject the button or toggle it automatically. The possibilities are almost endless and up to the add-on logic where to inject a context menu item or toggle automatically.
> 
> > "Current GUI settings / Player / Use HDR display capabilities" needs know if system supports HDR to shown setting.
> 
> And it still can. Nobody said that you have to remove display capability enumeration in KODI etc. - just the power menu button and the logic to change the system setting.
> 
> > Windows HDR switch from outside OK..... but swapchain is reconfigured by Python hahahhah
> 
> nobody said that everything had to move into an add-on - only the logic to flip the HDR switch in the OS. Yes, that would mean that some logic might have to be duplicated in the tool (HDR capability detection) or that add-ons might have to be able to read some system information via APIs from KODI. In fact, there are already ways to get system info from within the Python API or JSON-RPC API. Those APIs could be extended to provide additional info and maybe even a method to enable/disable KODIs HDR mode if this would be needed (like it's possible to trigger the stereoscopic 3D mode from python and JSON-RPC).
> So there would be ways to do what we asked for, but you completely shut down and don't seem to be open to compromise. That's really sad to see.
> 
> @dvidebaek we do want HDR, just not the power menu switch and the logic to change the OS setting. Everything else was fine (with room for later improvement), but thexai unfortunately does not seem to be willing to find a compromise to outsource this settings toggle or communicate with us in a cooperative way.

You are Right. Cooperation is the way forward. This should not be a one man show. Have you reached out to @thexai in private? Maybe TK and thexai can reach an mutual understading??

**Comment by chewitt** [comment]
I'm unable to comment on the PR content (not a Windows dev, can't spell SDK) but I'd observe that developers and maintainers have similar but different goals. Developers like to merge a new feature early then iterate it until everything is good. Maintainers prefer the development and iteration to be done first, then a finished feature gets merged. Release Managers are always balancing the needs of those two personalities. Early in a development cycle we can be gung-ho and many times we've borked Kodi for a while as something disruptive got merged, but as the release cycle evolves to public releases we need to behave more like maintainers. 

RIght now we're already overdue for shipping K19 Alpha1, so we're definitely in maintainer mode and will want things to be fully-baked before we let code merge. Also, and this is a critical point, requiring code to be effectively feature-complete before merge is the only sensible option when a contributor attempts to blackmail us with threats of stopping work. As maintainers, why would we risk incomplete features being merged? It's always sad that people attempt this behaviour as it's counter-productive and triggers the team's sense of responsibility for the codebase. Kodi is far from perfect (way too many chefs in the kitchen over too many years for that) but we know that, and from experience we know we need to be careful with it.

Anyway.. the door is never closed to any developer having a change of heart and making the effort to submit code. HDR for Windows is definitely a desired feature, so even if @thexai is not the right person some positive progress was made here and someone else can pick up the batton. At this stage it looks pretty unlikely that anything can/will be done for K19, but it's often best to let larger features land without being forced. It looks like we have decided K19 is the last release to support Win7, so K20 will avoid some issues there. K20 is also the target for several other platforms to have HDR support squared away. Again, the door is never closed.

**Comment by ciukacz2** [comment]
what is TK's stance on someone taking this exact PR's source code (well its GPLv2 after all) and resubmitting it with modifications at a later date?

**Comment by chewitt** [comment]
No objections as long as the original commit attribution is retained where possible/sensible in the revised submission. It's always preferable for the original contributor to finish the task, but if that's not to be - the beauty of open-source is - someone else can always continue the work.

**Comment by thexai** [comment]
I really don't understand how a simple button on a site that don't like makes life so complicated.

Collaboration is:

- Gets merged as is.
- Someone with skinning skills takes master branch and changes the toggle button from site/appearance.
- END

Why is need move HDR detection/toggle logic to external .exe   (trojan)  ??
Why is need a Python 3 addon to switch HDR?
Why?

**Comment by thexai** [comment]
I have also made a great effort to improve Win32 API detection/toggle code.

Even others have helped me in it.

There are no "magic bytes" anymore:


`if (ERROR_SUCCESS == GetDisplayConfigBufferSizes(QDC_ONLY_ACTIVE_PATHS, &pathCount, &modeCount))
  {
    std::vector<DISPLAYCONFIG_PATH_INFO> paths(pathCount);
    std::vector<DISPLAYCONFIG_MODE_INFO> modes(modeCount);

    if (ERROR_SUCCESS == QueryDisplayConfig(QDC_ONLY_ACTIVE_PATHS, &pathCount, paths.data(),
                                            &modeCount, modes.data(), 0))
    {
      DISPLAYCONFIG_GET_ADVANCED_COLOR_INFO getColorInfo = {};
      getColorInfo.header.type = DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO;
      getColorInfo.header.size = sizeof(getColorInfo);

      for (const auto& mode : modes)
      {
        if (mode.infoType == DISPLAYCONFIG_MODE_INFO_TYPE_TARGET)
        {
          getColorInfo.header.adapterId.HighPart = mode.adapterId.HighPart;
          getColorInfo.header.adapterId.LowPart = mode.adapterId.LowPart;
          getColorInfo.header.id = mode.id;
          if (ERROR_SUCCESS == DisplayConfigGetDeviceInfo(&getColorInfo.header))
          {
            if (getColorInfo.advancedColorEnabled)
              advancedColorEnabled = true;

            if (getColorInfo.advancedColorSupported)
              advancedColorSupported = true;
          }
        }
      }

      std::string txStatus;

      if (!advancedColorSupported)
      {
        status = 0;
        txStatus = "Display not HDR capable";
      }
      else if (advancedColorSupported && !advancedColorEnabled)
      {
        status = 1;
        txStatus = "Display HDR capable and current HDR status is disabled";
      }
      else if (advancedColorSupported && advancedColorEnabled)
      {
        status = 2;
        txStatus = "Display HDR capable and current HDR status is enabled";
      }

      // Prevents logging at application start before LOGLEVEL gets configured
      if (CServiceBroker::IsServiceManagerUp())
      {
        CLog::Log(LOGDEBUG, "{0:s} (status = {1:d})", txStatus, status);
      }
    }
  }

  return status;


Why is this code bad?
Why?

**Comment by thexai** [comment]
And yes, if in the end are going to do things wrong I prefer someone else to do it. 
(and stay out)

**Comment by Paxxi** [comment]
Locking this now. I wanted to give thexai a chance to respond before doing so.

Further discussion here will by the looks of it not improve things and is likely detrimental.

**Comment by da-anda** [comment]
I am not a Windows dev thexai, but the only thing I personally can spot that is "wrong" with your code above is that it uses integers instead of ENUMs as state values. We prefer ENUMs since they improve readability of the code, especially for devs not familiar with it, and the code is ofc a little less error prone.
